modcc/solvers.cpp

```diff
@@ -499,6 +536,43 @@ void LinearSolverVisitor::finalize() {
         statements_.push_back(std::move(local_t_term.assignment));
     }
 
+    // If size of matrix > 5, normalize intermediate variables
+    if (A_.size() > 5) {
+        for (unsigned s = 0; s < symbol_part.size() - 1; ++s) {
+            // Collect the elements in a single row to normalize
+            std::vector<expression_ptr> row_elems;
+            for (auto i = symbol_part[s]; i < symbol_part[s + 1]; ++i) {
+                row_elems.emplace_back(make_expression<IdentifierExpression>(Location(), name(symtbl_[i])));
+            }
+
+            // Get the max element in the row
+            expression_ptr max;
+            for (auto &elem: row_elems) {
+                auto abs = make_expression<AbsUnaryExpression>(elem->location(), elem->is_identifier()->clone());
+                if (!max) {
+                    max = std::move(abs);
+                } else {
+                    max = make_expression<MaxBinaryExpression>(elem->location(), max->clone(), std::move(abs));
+                }
+            }
+            // Safely inverse max
+            auto inv = make_expression<SafeInvUnaryExpression>(max->location(), std::move(max));
+
+            // Create a symbol for inv
+            auto local_inv_term = make_unique_local_assign(block_scope_, inv.get(), "inv_");
+
+            statements_.push_back(std::move(local_inv_term.local_decl));
+            statements_.push_back(std::move(local_inv_term.assignment));
+
+            // Update the row elements
+            for (auto &elem: row_elems) {
+                auto ratio = make_expression<MulBinaryExpression>(elem->location(), elem->clone(), std::move(local_inv_term.id));
+                auto assign = make_expression<AssignmentExpression>(elem->location(), std::move(elem), std::move(ratio));
+                statements_.push_back(std::move(assign));
+            }
+        }
+    }
+
```

Comment:
Now that the Gaussian elimination, divide and assign code in the finalize() methods for  LinearSolverVisitor,  SparseSolverVisitor and SparseNonlinearSolverVisitor are getting quite large, and share a lot of common code, could that code be factored out into its own routine?

Reply:
I was thinking about how to do that. Let me know what you think of the implementation.